contracts/NonFungibleToken-v2.cdc

```diff
@@ -0,0 +1,158 @@
+/**
+
+## The Flow Non-Fungible Token standard
+
+## `NonFungibleToken` contract interface
+
+The interface that all Non-Fungible Token contracts could conform to.
+If a user wants to deploy a new NFT contract, their contract would need
+to implement the NonFungibleToken interface.
+
+Their contract would have to follow all the rules and naming
+that the interface specifies.
+
+## `NFT` resource
+
+The core resource type that represents an NFT in the smart contract.
+
+## `Collection` Resource
+
+The resource that stores a user's NFT collection.
+It includes a few functions to allow the owner to easily
+move tokens in and out of the collection.
+
+## `Provider` and `Receiver` resource interfaces
+
+These interfaces declare functions with some pre and post conditions
+that require the Collection to follow certain naming and behavior standards.
+
+They are separate because it gives the user the ability to share a reference
+to their Collection that only exposes the fields and functions in one or more
+of the interfaces. It also gives users the ability to make custom resources
+that implement these interfaces to do various things with the tokens.
+
+By using resources and interfaces, users of NFT smart contracts can send
+and receive tokens peer-to-peer, without having to interact with a central ledger
+smart contract.
+
+To send an NFT to another user, a user would simply withdraw the NFT
+from their Collection, then call the deposit function on another user's
+Collection to complete the transfer.
+
+*/
+
+import MetadataViews from "./MetadataViews.cdc"
+
+/// The main NFT contract interface. Other NFT contracts will
+/// import and implement this interface
+///
+pub contract NonFungibleToken {
+
+    /// Interface that the NFTs have to conform to
+    ///
+    pub resource interface NFT { //: MetadataViews.Resolver {
+        /// The unique ID that each NFT has
+        pub let id: UInt64
+
+        pub fun getViews(): [Type]
+        pub fun resolveView(_ view: Type): AnyStruct?
+    }
+
+    /// Interface to mediate withdraws from the Collection
+    ///
+    pub resource interface Provider {
+        /// withdraw removes an NFT from the collection and moves it to the caller
+        pub fun withdraw(withdrawID: UInt64): @AnyResource{NFT} {
+            post {
+                result.id == withdrawID: "The ID of the withdrawn token must be the same as the requested ID"
+            }
```

Comment:
I don't have evidence, but have a strong feeling this will not be enough. We normally had type guarantees when adding NFT to collection ( benefit of single type ), this feels like weak guarantee here without type

Reply:
Good point! Do you have any suggestions for how we might enforce that?

Reply:
I don't have unfortunately, but I will think on this

Reply:
+1 on this, a malicious contract can pretty easily get around this unfortunately. If we did it based on uuid that would be a different story but I don't see a good way to make that happen

Reply:
I meant to have `getID()` as a function instead of requiring the ID field. I've updated the code to use that now. Thanks for pointing it out!

Reply:
`getID()` has very bad synergy with this, post condition losing all meaning.

Reply:
> getID() has very bad synergy with this, post condition losing all meaning. 

I see two ways this could cause a weak post condition:

1. The post condition doesn't check the type of the NFT. A user may think they're going to get TopShot 42 but actually get TopShotPack 42. Is this what you're worried about @bluesign?
2. The `getID` function is dynamic and could return a different value each time it's called. This feels like less of a problem but still prevents the post condition from giving strong guarantee.
    - I'm actually more worried about this in the general case. I think we probably want IDs to be constant identifiers of an NFT. Collections only work under this assumption. But making the ID a function could encourage developers to dynamically generate their ID based on data that could change, which feels fragile to me.